cranelift/isle/isle/src/trie_again.rs

```diff
@@ -0,0 +1,543 @@
+        /// we track them by index for constant-time comparisons. The [sema::TypeEnv] can be used to
+        /// get the field names.
+        field: TupleIndex,
+    },
+    /// After some sequence of matches, we'll match one of the previous bindings against
+    /// `Option::Some` and produce a new binding from its contents. (This currently only happens
+    /// with external extractors.)
+    Some {
+        /// Which binding is being matched?
+        source: BindingId,
+    },
+    /// After some sequence of matches, we'll match one of the previous bindings against a tuple and
+    /// produce a new binding from one of its fields. (This currently only happens with external
+    /// extractors.)
+    Tuple {
+        /// Which binding is being matched?
+        source: BindingId,
+        /// Which tuple field are we projecting out?
+        field: TupleIndex,
+    },
+}
+
+/// Pattern matches which can fail.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Constraint {
+    /// The value must match this enum variant.
+    Variant(TupleIndex, sema::TypeId, sema::VariantId),
+    /// The value must equal this integer literal.
+    ConstInt(i128),
+    /// The value must equal this Rust primitive value.
+    ConstPrim(sema::Sym),
+    /// The value must be an `Option::Some`, from a fallible extractor.
+    Some,
+}
+
+/// Expressions construct new values. Rust pattern matching can only destructure existing values,
+/// not call functions or construct new values. So `if-let` and external extractor invocations need
+/// to interrupt pattern matching in order to evaluate a suitable expression. These expressions are
+/// also used when evaluating the right-hand side of a rule.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub enum Expr {
+    /// Evaluates to the given integer literal.
+    ConstInt(i128),
+    /// Evaluates to the given primitive Rust value.
+    ConstPrim(sema::Sym),
+    /// One of the arguments to the top-level function.
+    Argument(TupleIndex),
+    /// A binding from some sequence of pattern matches.
+    Binding(BindingId),
+    /// The result of calling an external extractor.
+    Extractor(sema::TermId, ExprId),
+    /// The result of constructing an enum variant.
+    Variant(sema::TypeId, sema::VariantId, Box<[ExprId]>),
+    /// The result of calling an external constructor.
+    Constructor(sema::TermId, Box<[ExprId]>),
+}
+
+/// A term-rewriting rule. All [BindingId]s and [ExprId]s are only meaningful in the context of the
+/// [RuleSet] that contains this rule.
+#[derive(Debug, Default)]
+pub struct Rule {
+    /// Where was this rule defined?
+    pub pos: Pos,
+    /// All of these bindings must match for this rule to apply. Note that within a single rule, if
+    /// a binding site must match two different constants, then the rule can never match.
+    pub constraints: HashMap<BindingId, Constraint>,
```

Comment:
`BTreeMap` has the disadvantage of being slower to compile than `HashMap` due to generating a lot more LLVM ir. cranelift-isle is decently fast to compule, so it doesn't matter that much right now, but it might be something to keep in mind.

Reply:
The way I'm expecting to use this is to iterate over `RuleSet::bindings` and then look up constraints accordingly. I guess a public `get` method with a private `HashMap` should be fine for that...